src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -1803,14 +1812,51 @@ bool LibraryCallKit::inline_vector_rearrange() {
   ciKlass* shbox_klass = shuffle_klass->const_oop()->as_instance()->java_lang_Class_klass();
   const TypeInstPtr* shbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, shbox_klass);
 
-  Node* v1 = unbox_vector(argument(4), vbox_type, elem_bt, num_elem);
-  Node* shuffle = unbox_vector(argument(5), shbox_type, shuffle_bt, num_elem);
+  ciKlass* mbox_klass = mask_klass->const_oop()->as_instance()->java_lang_Class_klass();
+  const TypeInstPtr* mbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, mbox_klass);
+
+  Node* v1 = unbox_vector(argument(5), vbox_type, elem_bt, num_elem);
+  Node* shuffle = unbox_vector(argument(6), shbox_type, shuffle_bt, num_elem);
+
+  bool is_masked_op = argument(7)->bottom_type() != TypePtr::NULL_PTR;
+  Node* mask = is_masked_op ? unbox_vector(argument(7), mbox_type, elem_bt, num_elem) : NULL;
+  if (is_masked_op && mask == NULL) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not supported: arity=3 op=shuffle/rearrange vlen=%d etype=%s ismask=useload is_masked_op=1",
+                    num_elem, type2name(elem_bt));
+    }
+    return false;
+  }
+
+  bool use_predicate = is_masked_op && arch_supports_vector(Op_VectorRearrange, num_elem, elem_bt, VecMaskUsePred);
+  if (is_masked_op && !use_predicate &&
+       (!arch_supports_vector(Op_VectorBlend, num_elem, elem_bt, VecMaskUseLoad) ||
+        !arch_supports_vector(VectorNode::replicate_opcode(elem_bt), num_elem, elem_bt, VecMaskNotUsed))) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not supported: arity=3 op=shuffle/rearrange vlen=%d etype=%s ismask=useload is_masked_op=1",
+                    num_elem, type2name(elem_bt));
+    }
+    return false;
+  }
 
-  if (v1 == NULL || shuffle == NULL) {
+  if (v1 == NULL || shuffle == NULL || mask == NULL) {
```

Comment:
`mask` is `NULL` for non-masked rearrange, right? I guess this might make the non-masked rearrange intrinsify failed?